File: app/src/main/java/ru/cj264/geekbrains/android_intro/homework/ui/notes_list/NotesListViewModel.java
package ru.cj264.geekbrains.android_intro.homework.ui.notes_list;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import ru.cj264.geekbrains.android_intro.homework.domain.Note;
import ru.cj264.geekbrains.android_intro.homework.domain.NotesRepository;
import ru.cj264.geekbrains.android_intro.homework.ui.notes_list.adapter.AdapterItem;
import ru.cj264.geekbrains.android_intro.homework.ui.notes_list.adapter.NoteAdapterItem;

public class NotesListViewModel extends ViewModel {
    private final NotesRepository repository;

    private final MutableLiveData<ArrayList<Note>> notesListLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> progressLiveData = new MutableLiveData<>();

    public NotesListViewModel(NotesRepository notesRepository) {
        repository = notesRepository;
    }

    public void fetchNotes() {
        progressLiveData.setValue(true);
        repository.getNotes(value -> {
            notesListLiveData.setValue(value);
            progressLiveData.setValue(false);
        });
    }

    public LiveData<List<AdapterItem>> getNotesListLiveData() {
        return Transformations.map(notesListLiveData, input -> {
            ArrayList<AdapterItem> output = new ArrayList<>();

            Collections.sort(input, (o1, o2)
                    -> o1.getCreationDateTime().compareTo(o2.getCreationDateTime()));

            for (Note note : input) {
                output.add(new NoteAdapterItem(note));
            }

            return output;
        });
    }

    public LiveData<Boolean> getProgressLiveData() {
        return progressLiveData;
    }

    public void addNewNote() {
        progressLiveData.setValue(true);
        repository.addNewNote(value -> {
            ArrayList<Note> currentNotes = notesListLiveData.getValue();
            currentNotes.add(value);
            notesListLiveData.postValue(currentNotes);
            progressLiveData.setValue(false);
        });
    }

    public void clearNotes() {
        progressLiveData.setValue(true);
        repository.clearNotes(voidValue -> {
            // TODO: clear repository? Danger!
            notesListLiveData.postValue(new ArrayList<>());
            progressLiveData.setValue(false);
        });
    }

    public void deleteAtPosition(int contextMenuItemPosition) {
        progressLiveData.setValue(true);
        // TODO: repository delete
        repository.addNewNote(value -> {
            ArrayList<Note> currentNotes = new ArrayList<>(notesListLiveData.getValue());
            currentNotes.remove(contextMenuItemPosition);
            notesListLiveData.postValue(currentNotes);
            progressLiveData.setValue(false);
        });
    }

}

File: app/src/main/java/ru/cj264/geekbrains/android_intro/homework/ui/notes_list/NotesListViewModel.kt
package ru.cj264.geekbrains.android_intro.homework.ui.notes_list

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import ru.cj264.geekbrains.android_intro.homework.domain.Note
import ru.cj264.geekbrains.android_intro.homework.domain.NotesRepository
import ru.cj264.geekbrains.android_intro.homework.ui.notes_list.adapter.AdapterItem
import ru.cj264.geekbrains.android_intro.homework.ui.notes_list.adapter.NoteAdapterItem
import java.util.*

class NotesListViewModel(private val repository: NotesRepository) : ViewModel() {
    private val notesListLiveData = MutableLiveData<ArrayList<Note>>()
    private val progressLiveData = MutableLiveData<Boolean>()

    fun fetchNotes() {
        progressLiveData.value = true
        repository.getNotes { value: ArrayList<Note> ->
            notesListLiveData.value = value
            progressLiveData.setValue(false)
        }
    }

    fun getNotesListLiveData(): LiveData<List<AdapterItem>> {
        return Transformations.map(notesListLiveData) { input: ArrayList<Note> ->
            val output = ArrayList<AdapterItem>()
            input.sortWith { o1, o2 -> o1.creationDateTime.compareTo(o2.creationDateTime) }
            for (note in input) {
                output.add(NoteAdapterItem(note))
            }
            output
        }
    }

    fun getProgressLiveData(): LiveData<Boolean> {
        return progressLiveData
    }

    fun addNewNote() {
        progressLiveData.value = true
        repository.addNewNote { value: Note ->
            val currentNotes = notesListLiveData.value!!
            currentNotes.add(value)
            notesListLiveData.postValue(currentNotes)
            progressLiveData.setValue(false)
        }
    }

    fun clearNotes() {
        progressLiveData.value = true
        repository.clearNotes { voidValue: Void? ->
            // TODO: clear repository? Danger!
            notesListLiveData.postValue(ArrayList())
            progressLiveData.setValue(false)
        }
    }

    fun deleteAtPosition(contextMenuItemPosition: Int) {
        progressLiveData.value = true
        // TODO: repository delete
        repository.addNewNote { value: Note? ->
            val currentNotes = ArrayList(notesListLiveData.value)
            currentNotes.removeAt(contextMenuItemPosition)
            notesListLiveData.postValue(currentNotes)
            progressLiveData.setValue(false)
        }
    }
}